crates/nu-lsp/src/symbols.rs

```diff
@@ -0,0 +1,534 @@
+            cache: BTreeMap::new(),
+            dirty_flags: BTreeMap::new(),
+        }
+    }
+
+    pub fn mark_dirty(&mut self, uri: Uri, flag: bool) {
+        self.dirty_flags.insert(uri, flag);
+    }
+
+    fn get_symbol_by_id(
+        working_set: &StateWorkingSet,
+        id: Id,
+        doc: &FullTextDocument,
+        doc_span: &Span,
+    ) -> Option<Symbol> {
+        match id {
+            Id::Declaration(decl_id) => {
+                let decl = working_set.get_decl(decl_id);
+                let span = working_set.get_block(decl.block_id()?).span?;
+                // multi-doc working_set, returns None if the Id is in other files
+                if !doc_span.contains(span.start) {
+                    return None;
+                }
+                Some(Symbol {
+                    name: decl.name().to_string(),
+                    kind: SymbolKind::FUNCTION,
+                    range: span_to_range(&span, doc, doc_span.start),
+                })
+            }
+            Id::Variable(var_id) => {
+                let var = working_set.get_variable(var_id);
+                let span = var.declaration_span;
+                if !doc_span.contains(span.start) || span.end == span.start {
+                    return None;
+                }
+                let range = span_to_range(&span, doc, doc_span.start);
+                let name = doc.get_content(Some(range));
+                // TODO: better way to filter closures with type any
+                if name.contains('\r')
+                    || name.contains('\n')
+                    || name.contains('{')
+                    || var.ty == Type::Closure
+                    || var.ty == Type::Block
+                {
+                    return None;
+                }
+                Some(Symbol {
+                    name: name.to_string(),
+                    kind: SymbolKind::VARIABLE,
+                    range,
+                })
+            }
+            _ => None,
+        }
+    }
+
+    fn extract_all_symbols(
+        working_set: &StateWorkingSet,
+        doc: &FullTextDocument,
+        cached_file: &CachedFile,
+    ) -> Vec<Symbol> {
+        let mut all_symbols: Vec<Symbol> = (0..working_set.num_decls())
+            .filter_map(|id| {
+                Self::get_symbol_by_id(
+                    working_set,
+                    Id::Declaration(DeclId::new(id)),
+                    doc,
+                    &cached_file.covered_span,
+                )
+            })
+            .chain((0..working_set.num_vars()).filter_map(|id| {
+                Self::get_symbol_by_id(
+                    working_set,
+                    Id::Variable(VarId::new(id)),
+                    doc,
+                    &cached_file.covered_span,
+                )
+            }))
+            // TODO: same variable symbol can be duplicated with different VarId
+            .collect::<HashSet<Symbol>>()
+            .into_iter()
+            .collect();
```

Comment:
Is an identical variable (same in terms of its span) supposed to have different VarId s?

Reply:
I haven't checked the cause of the duplication, just observed that we can get the same span with `working_set.get_var` on different IDs. So I manually clean them in current implementation.

Reply:
I did some hacking. I'm not sure why there seem to be dupes.
![image](https://github.com/user-attachments/assets/e1fde07a-74b1-4281-aead-c38447a081e2)